keps/sig-node/0034-20190129-pid-limiting.md

```diff
@@ -0,0 +1,186 @@
+restricted and/or defaulted via PodSecurityPolicy or LimitRange.  We anticipate
+tracking this work under a separate feature gate `GranularPidLimitsPerPod`.  Any
+defaulting applied to pods today would only be used if the pod had no local pod
+pid limiting policy in future dates.
+
+## Proposal
+
+### User Stories [optional]
+
+1. Administrator can default the number of pids per pod to provide pod-to-pod
+   isolation.
+1. Administrator can reserve a number of allocatable pids to user pods via node
+   allocatable.
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### Pod to Pod Isolation
+
+To enable pid isolation among pods, the `SupportPodPidsLimit` feature gate is
+defined.
+
+If enabled, the kubelet argument for `pod-max-pids` will write out the
+configured pid limit to the pod level cgroup to the value specified on Linux
+hosts.  If -1, the kubelet will default to the node allocatable pid capacity.
+
+#### Node to Pod Isolation
+
+To enable pid isolation from node to pods, the `SupportNodePidsLimit` feature
+gate is proposed.  If enabled, pid reservations may be supported at the node
+allocatable and eviction manager subsystem configurations.
+
+Node allocatable is a well-established feature concept in the kubelet that
+allows isolation of user pod resources from host daemons at the `kubepods`
+cgroup level that parents all end-user pods.
+
+The kubelet will be updated to support reservation of pids so the effective pid
+limit is enabled as follows:
+
+```
+[Allocatable] = [Node Capacity] - 
+ [Kube-Reserved] - 
+ [System-Reserved] - 
+ [Hard-Eviction-Threshold]
+```
+
+#### Cgroup Enforcement
+
+To use this feature, the `--cgroups-per-qos` must be enabled.  In addition, the
+`pids` cgroup must be mounted.
+
+The `kubepods` cgroup is bounded by the `Allocatable` value.
+
+The QoS level cgroups are left unbounded across all pid pool sizes.
+
+The pod level cgroup sandbox is configured as follows:
+
+1. the pod-max-pids value if positive and is specified on kubelet config
+1. the local pod pid limiting policy (future)
+1. unbounded (so it is restricted by the `Allocatable` value at `kubepods`)
+
+### Risks and Mitigations
+
+None
+
+## Graduation Criteria
+
+### Pod to Pod pid isolation
+
+The following criteria applies to `SupportPodPidsLimit` feature gate:
+
+Alpha
+- basic support integrated in kubelet
+
+Beta
+- ensure proper node e2e test coverage is integrated verifying cgroup settings
+- see testing:
+https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/pids_test.go
+
+GA
+- assuming no negative user feedback based on production experience, promote
+  after 2 releases in beta.
+
+### Node to Pod pid isolation
+
+Adding support for pid limiting at the Node Allocatable level 
+
+The following criteria applies to `SupportNodePidsLimit`:
+
+Alpha
+- basic support integrated via eviction manager and/or node allocatable level
+
+Beta
+- ensure proper node e2e testing coverage to ensure a pod is unable to fork-bomb
```

Comment:
You can link to the fork-bomb test I wrote here: https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/eviction_test.go#L397.
Although, we should ideally have separate tests for node-level thresholds and allocatable-level thresholds.